extensions/caffeine-cache/src/main/java/io/druid/client/cache/CaffeineCache.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.client.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import com.google.common.primitives.Ints;
+import com.metamx.common.logger.Logger;
+import com.metamx.emitter.service.ServiceEmitter;
+import com.metamx.emitter.service.ServiceMetricEvent;
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4Factory;
+import org.apache.commons.codec.digest.DigestUtils;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class CaffeineCache implements io.druid.client.cache.Cache
+{
+  private static final Logger log = new Logger(CaffeineCache.class);
+  private final Cache<String, byte[]> cache;
+  private final AtomicReference<com.github.benmanes.caffeine.cache.stats.CacheStats> priorStats = new AtomicReference<>(
+      null);
+
+
+  public static CaffeineCache create(final CaffeineCacheConfig config)
+  {
+    Caffeine<Object, Object> builder = Caffeine.newBuilder().recordStats();
+    if (config.getExpiration() >= 0) {
+      builder = builder
+          .expireAfterAccess(config.getExpiration(), TimeUnit.MILLISECONDS);
+    }
+    if (config.getMaxSize() >= 0) {
+      builder = builder
+          .maximumSize(config.getMaxSize());
+    }
+    return new CaffeineCache(builder.<String, byte[]>build());
```

Comment:
#1849 and #2351 both use `maximumWeight` with a weigher, so I suspect you may want to do similar. I'm also a little disappointed that Java didn't infer the type for you.